src/components/TermSearchControl/index.js

```diff
@@ -0,0 +1,185 @@
+		},
+		[ taxonomy ]
+	);
+	const { searchResults, searchHasResolved } = useSelect(
+		( select ) => {
+			if ( ! search ) {
+				return {
+					searchResults: EMPTY_ARRAY,
+					searchHasResolved: true,
+				};
+			}
+			const { getEntityRecords, hasFinishedResolution } = select( coreStore );
+			const selectorArgs = [
+				'taxonomy',
+				taxonomy,
+				{
+					...BASE_QUERY,
+					search,
+					orderby: 'name',
+					exclude: termIds,
+					per_page: 20,
+				},
+			];
+			return {
+				searchResults: getEntityRecords( ...selectorArgs ),
+				searchHasResolved: hasFinishedResolution(
+					'getEntityRecords',
+					selectorArgs
+				),
+			};
+		},
+		[ search, termIds ]
+	);
+
+	// `existingTerms` are the ones fetched from the API and their type is `{ id: number; name: string }`.
+	// They are used to extract the terms' names to populate the `FormTokenField` properly
+	// and to sanitize the provided `termIds`, by setting only the ones that exist.
+	const existingTerms = useSelect(
+		( select ) => {
+			if ( ! termIds?.length ) {
+				return EMPTY_ARRAY;
+			}
+			const { getEntityRecords } = select( coreStore );
+			return getEntityRecords( 'taxonomy', taxonomy, {
+				...BASE_QUERY,
+				include: termIds,
+				per_page: termIds.length,
+			} );
+		},
+		[ termIds ]
+	);
+
+	// Update the `value` state only after the selectors are resolved
+	// to avoid emptying the input when we're changing terms.
+	useEffect( () => {
+		if ( ! termIds?.length ) {
+			setValue( EMPTY_ARRAY );
+		}
+		if ( ! existingTerms?.length ) {
+			return;
+		}
+		// Returns only the existing entity ids. This prevents the component
+		// from crashing in the editor, when non existing ids are provided.
+		const sanitizedValue = termIds.reduce( ( accumulator, id ) => {
+			const entity = existingTerms.find( ( term ) => term.id === id );
+			if ( entity ) {
+				accumulator.push( {
+					id,
+					value: entity.name,
+				} );
+			}
+			return accumulator;
+		}, [] );
+		setValue( sanitizedValue );
+	}, [ termIds, existingTerms ] );
+
+	// Update suggestions only when the query has resolved.
+	useEffect( () => {
+		if ( ! searchHasResolved ) {
+			return;
+		}
+		setSuggestions( searchResults.map( ( result ) => result.name ) );
+	}, [ searchResults, searchHasResolved ] );
+
+	/**
+	 * Function to handle change of selected terms.
+	 *
+	 * @param {Array} newTermValues Array of new term values.
+	 */
+	const onTermsChange = ( newTermValues ) => {
+		const newTermIds = new Set();
+		for ( const termValue of newTermValues ) {
+			const termId = getTermIdByTermValue( searchResults, termValue );
+			if ( termId ) {
+				newTermIds.add( termId );
+			}
+		}
+		setSuggestions( EMPTY_ARRAY );
+		onChange( Array.from( newTermIds ) );
+	};
+
+	return (
+		<FormTokenField
+			displayTransform={ decodeEntities }
+			label={
+				label ||
+				sprintf(
+					__( 'Filter by %s', 'block-editor-components' ),
+					taxObject ? taxObject?.labels?.singular_name : ''
```

Comment:
Does the fallback to an empty string make sense? I would think something like `__( 'term' )` would be a reasonable fallback so the UI doesn't show an incomplete phrase.